### PR TITLE
[WIP] History feature on CLI

### PR DESCRIPTION
**Pull request by guilherssousa**
This PR introduces the foundations of the new History feature.

- [x] Create a History store to hold information about recent actions
- [x] Add an Interface to navigate between past issues
- [x] Preprocess actions to correctly showcase replacements

### CODE REVIEW

File: src/display.ts [hunk 1]
```diff
@@ -119,6 +127,60 @@ export const determineAction = async (
 	return [action, replacer];
 };
 
+export const determineHistoryIssue = async () => {
```
Comment (Lioness100) [inline src/display.ts:130]: I think this function should be renamed, as well as `HistoryIssue`. Doesn't sound right to me. Maybe `PastIssue` or something?

File: src/handleIssue.ts [hunk 2]
```diff
@@ -160,6 +161,28 @@ export const handleIssue = async (issues: Issue[], issue: Issue) => {
...
+
+		const hasQuit = await handleIssue(issues, issue);
+
+		if (hasQuit) {
+			return true;
+		}
+
+		removeIssueFromHistory(id);
+
+		return handleIssues(issues);
```
Comment (Lioness100) [inline src/handleIssue.ts:183]: This won't work for a few reasons, one of which is that `handleIssues` stores the length of the issues in a variable to be displayed in the header of the action selector. This would rewrite the length. Look into the existing implementation of undo last action to see how I did it.

File: src/history.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
+import { randomUUID } from 'node:crypto';
+import { type HistoryIssue } from './shared';
+
+export const history = new Array<HistoryIssue>();
+
+export const addNewHistoryIssue = ({ action, issue, replacer }: Omit<HistoryIssue, 'id'>) => {
+	if (history.length >= 100) {
+		history.shift();
+	}
```
Comment (Lioness100) [inline src/history.ts:9]: 100 concerns me because I would think that would totally flood the user's console? Also this wouldn't really alleviate the problem we discussed over discord since it's still a very large amount of issues. I was thinking 5 to 10. If we do 100, maybe consider pagination?

File: src/history.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+	let original: string;
+
+	//  If action is Replace or ReplaceAll, the word should be replaced with the replacer
+	// on the issue object, because when the user rolls back the action and write a replacement
+	// again, the replacer should be the new word to replace the issue to avoid overflowing characters.
+	if (replacer) {
+		// Don't record the action if the replacer is the same as the issue text.
+		if (replacer === issue.text) {
```
Comment (Lioness100) [inline src/history.ts:18]: This is a good idea, could you add this check in display.ts too where it is like `if (!replacer) { return Action.Ignore };`

File: src/history.ts [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+			return;
+		}
+
+		original = issue.text;
+
+		issue.line.text = issue.line.text.replace(issue.text, replacer);
```
Comment (Lioness100) [inline src/history.ts:24]: This will have to be more complicated and use offsets. For example, what if the same text is in the line twice and the user ignores the first occurrence? See existing mplementation for hints.

File: src/history.ts [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+		issue.text = replacer;
+		issue.length = replacer.length;
```
Comment (Lioness100) [inline src/history.ts:27]: This property isn't used

File: src/history.ts [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+	}
+
+	history.push({
+		action,
+		issue,
+		original: original!,
+		replacer,
+		id: randomUUID()
```
Comment (Lioness100) [inline src/history.ts:35]: I don't understand why we need UUID's, why don't you just compare objects?

File: src/shared.ts [hunk 8]
```diff
@@ -2,13 +2,22 @@ import type { CSpellSettings, Issue } from 'cspell';
...
-	UndoLastAction,
-	Quit
+	Ignore = 'ignore',
+	IgnoreAll = 'ignore-all',
+	OpenHistory = 'open-history',
+	Quit = 'quit',
+	Replace = 'replace',
+	ReplaceAll = 'replace-all',
+	SkipFile = 'skip-file',
+	UndoLastAction = 'undo-last-action'
```
Comment (Lioness100) [inline src/shared.ts:12]: Why did you add string values?

File: src/history.ts [hunk 9]
```diff
@@ -0,0 +1,45 @@
+import { randomUUID } from 'node:crypto';
+import { type HistoryIssue } from './shared';
+
+export const history = new Array<HistoryIssue>();
```
Comment (Lioness100) [inline src/history.ts:4]:
```suggestion
export const history: HistoryIssue[] = [];
```

**Review by Lioness100 — CHANGES_REQUESTED** [review]
Thanks so much for the pr! Lmk if you need any help with my requested changes.

File: src/shared.ts (hunk 8 above)
Comment (guilherssousa) [reply src/shared.ts:12]: Was easier to debug when logging 😬 I'll remove this on final commits

Comment (Lioness100) [reply src/shared.ts:12]: Ok, no problem :)

File: src/display.ts [hunk 10]
```diff
@@ -79,6 +79,11 @@ export const determineAction = async (
 	const otherTypoInstancesCount = issues.filter((otherIssue) => otherIssue.text === issue.text).length;
 	const otherTyposInFileCount = issues.filter((otherIssue) => otherIssue.uri === issue.uri).length;
 
+	// The explicit undefined check is needed because Action.Ignore is 0, which is falsy.
+	const shouldDisplayUndoLastAction =
+		previousState.action !== undefined && previousState.action !== Action.UndoLastAction;
+	const shouldDisplayOpenHistory = previousState.action !== undefined && history.length > 0;
```
Comment (Lioness100) [inline src/display.ts:85]: Wouldn't `history.length > 0` be implied if there was a previous action? Also, in general, should we remove most of `previousState` and just treat "undo last action" as the same as viewing the history and clicking the most recent action? This would reduce code duplication.

File: src/display.ts [hunk 11]
```diff
@@ -89,12 +94,13 @@ export const determineAction = async (
...
+				? [{ name: yellowBright('Undo Last Action'), value: Action.UndoLastAction }]
+				: []),
+			...(shouldDisplayOpenHistory
 				? [
-						{ name: yellowBright('Undo Last Action'), value: Action.UndoLastAction },
 						{
-							name: magentaBright('Open Issue History'),
+							name: magentaBright('Open Action History'),
 							value: Action.OpenHistory
 						}
```
Comment (Lioness100) [inline src/display.ts:105]: Nitpick, could this be formatted on one line like the "undo last action"?

File: src/history.ts (hunk 7 above)
Comment (guilherssousa) [reply src/history.ts:35]: On my frist implementation I used Maps instead of objects, but dealing with them during the listings for `detemineHistoryIssue` with methods like `.forEach` was a little problematic, so I changed to objects and now there's no reason to use UUIDs there, I'll remove it

File: src/history.ts [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+		issue.line.text = issue.line.text.replace(issue.text, replacer);
+
+		issue.text = replacer;
+		issue.length = replacer.length;
+	}
+
+	history.push({
+		action,
+		issue,
+		original: original!,
```
Comment (Lioness100) [inline src/history.ts:33]: This doesn't seem like it needs the non null assertion
```suggestion
		original,
```

Comment (guilherssousa) [reply src/history.ts:33]: Solved!

File: src/history.ts (hunk 3 above)
Comment (guilherssousa) [reply src/history.ts]: I replaced it with just 10 on last commit :)

File: src/handleIssue.ts [hunk 13]
```diff
@@ -187,7 +211,14 @@ export const handleIssue = async (issues: Issue[], issue: Issue) => {
 		}
 	}
 
+	addNewHistoryIssue({
+		action,
+		issue,
+		replacer
+	});
```
Comment (Lioness100) [inline src/handleIssue.ts:218]: Reformat to one line, please

File: src/history.ts [hunk 14]
```diff
@@ -0,0 +1,41 @@
+import { type HistoryIssue } from './shared';
+
+export const history: HistoryIssue[] = [];
+
+export const addNewHistoryIssue = ({ action, issue, replacer }: Omit<HistoryIssue, 'id'>) => {
+	if (history.length >= 10) {
+		history.shift();
+	}
+
+	const original: string = issue.text;
```
Comment (Lioness100) [inline src/history.ts:10]: Unneeded explicit type

File: src/history.ts [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+		issue,
+		original,
+		replacer
+	});
+};
+
+export const removeIssueFromHistory = (issue: HistoryIssue) => {
+	const index = history.indexOf(issue);
+
+	if (index >= 0) {
```
Comment (Lioness100) [inline src/history.ts:38]: I think !== -1 is better

File: src/history.ts [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+		issue.text = replacer;
+		issue.length = replacer.length;
+	}
+
+	history.push({
+		action,
+		issue,
+		original,
+		replacer
+	});
```
Comment (Lioness100) [inline src/history.ts:32]: One line formatting ditto

File: src/display.ts (hunk 11 above)
Comment (guilherssousa) [reply src/display.ts]: Done!

File: src/handleIssue.ts [hunk 17]
```diff
@@ -160,6 +161,28 @@ export const handleIssue = async (issues: Issue[], issue: Issue) => {
...
 	}
 
+	if (action === Action.OpenHistory) {
+		resetDisplay();
+
+		const determinedAction = await determineHistoryIssue();
+
+		if (determinedAction === Action.Quit) {
+			return true;
+		}
```
[the remark this answers is not in the document]
Comment (guilherssousa) [reply src/handleIssue.ts:172]: I changed this Quit value with just a undefined one, so the variable name stays more coerent to its function on `handleIssue`

File: src/display.ts [hunk 18]
```diff
@@ -119,6 +127,60 @@ export const determineAction = async (
...
+		name: 'issue',
+		message: 'Choose a issue to go back:',
+		choices: [
+			{ name: red('Quit'), value: Action.Quit },
+			...history.map((entry) => {
+				let name: string;
+
+				const uri = fileURLToPath(new URL(entry.issue.uri!));
+
+				switch (entry.action) {
```
[the remark this answers is not in the document]
Comment (guilherssousa) [reply src/display.ts:148]: `actionDisplays` can be declared outside the function, but still it needs arguments to concatenate the final string, right? So, to serve them outside of the function, defining another one to receive these arguments woudn't be redundant?

```ts
const actionDisplays: Partial<Record<Action, string>> = {
	[Action.Ignore]: `❕ Ignored ${cyan(entry.issue.text)}`,
	[Action.Replace]: `✏ Replaced ${cyan(entry.original!)} with ${cyan(entry.replacer!)}`,
	[Action.IgnoreAll]: `❕ Ignored all occurrences of ${cyan(entry.issue.text)}`,
	[Action.ReplaceAll]: `✏ Replaced all occurrences of ${cyan(entry.original!)}`
};

```

File: src/history.ts (hunk 4 above)
Comment (guilherssousa) [reply src/history.ts:17]: You mean this? 😬
```ts
if (!replacer || replacer === issue.text) {
  return [Action.Ignore, ''];
}
```

File: src/display.ts (hunk 18 above)
[the remark this answers is not in the document]
Comment (Lioness100) [reply src/display.ts:148]: Oh, you're right, it shouldn't be outside the function

File: src/history.ts (hunk 4 above)
Comment (Lioness100) [reply src/history.ts:17]: Yeah, I was thinking that if someone accidentally clicks the replace button, they might think, "oh, I'll just type in the same word so nothing changes".

Comment (guilherssousa) [reply src/history.ts:17]: You have a point, even I made it once 🤣

File: src/display.ts (hunk 10 above)
Comment (guilherssousa) [reply src/display.ts:85]: I pushed some changes I made on the past days recently, but some reviews weren't complete, like this one 🔝, I'll check it today

File: src/display.ts [hunk 19]
```diff
@@ -138,39 +138,17 @@ export const determineHistoryIssue = async () => {
...
-					}
-					default: {
-						prompt = 'Unknown action';
-					}
-				}
+				const actionDisplays: Partial<Record<Action, string>> = {
+					[Action.Ignore]: `❕ Ignored ${cyan(entry.issue.text)}`,
+					[Action.Replace]: `✏ Replaced ${cyan(entry.original!)} with ${cyan(entry.replacer!)}`,
+					[Action.IgnoreAll]: `❕ Ignored all occurrences of ${cyan(entry.issue.text)}`,
+					[Action.ReplaceAll]: `✏ Replaced all occurrences of ${cyan(entry.original!)}`
```
Comment (Lioness100) [inline src/display.ts:145]: Missing the "with ${cyan(entry.replacer!)}"

File: src/display.ts [hunk 20]
```diff
@@ -138,39 +138,17 @@ export const determineHistoryIssue = async () => {
...
+				};
 
 				return {
-					name: `${prompt} in ${cyan(uri)}:${cyan(entry.issue.row)}:${cyan(entry.issue.col)}`,
+					name: `${actionDisplays[entry.action] ?? 'Unknown action'} in ${cyan(uri)}:${cyan(
+						entry.issue.row
+					)}:${cyan(entry.issue.col)}`,
```
Comment (Lioness100) [inline src/display.ts:151]: Nitpick, can you separate the highlighted path to a variable and then have the return statement be one line?

**Review by Lioness100 — COMMENTED** [review]
Forgive me for being nitpicky and indecisive, I can make this change for you if you want. I still don't really think "past issue" is good wording. Maybe, `pickIssueFromHistory`, `appendIssueToHistory`, and `ActionHistoryEntry`?

**Comment by guilherssousa** [comment]
Hi, sorry for not updating this PR for so long, unfortunately, I won't be able to continue working on this feature, as I'm currently facing tight deadlines and demands both at work and in my studies. So I will have to prioritize those for some time and archive this contribution 😞 Feel free to continue ongoing works if you want or I can close the PR
